### DISCUSSION TITLE
reduce/fold/accumulate on large data objects

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#241] [Design & Development]
I have the guts worked out for a mspass implementation of pwmig.   For readers other than Ian Wang, pwmig is a compute intensive program Ian used for his dissertation.  A single run of the workflow he used for his phd dissertation took around 30 cpu days that became feasible only by a crude parallelization that reduced that to the order of a day of wall time.  

pwmig can be abstracted as 3 loops:
1.  An outer loop over events (the only level of parallelization in the original C++ code).
2. An secondary loop over "plane wave components" - a unique feature of this algorithm
3. A inner loop that works one "ray path" at a time.  

I've implemented maps functions to handle all the data up to the common point in the algorithm where all 3 loops truncate on what in spark would be called a reduce step and dask would call *accumulate* or *fold*.   

The fundamental thing I'm worried about and can't find any useful information on is how the reduce steps work with large memory returns.  All three reduce steps can be cast as sums of very large data object (well, not massive but pretty huge - several gigabytes).   I'm hoping within a single node that will be a pointer return, but if a reduce/accumulate/fold function has to be serialized for every single sum element this will be very very slow.   

Any insight anyone can provide on this?  We'll know when I get this to the point of working, but if I've facing a deep hole a redesign might be in order.

**Comment by wangyinz** [comment]
hmmm... if there is no way to further express the parallelism with smaller data object, probably the answer is that we shouldn't rely on spark or dask for this step. They were not designed to efficiently handle such an algorithm as this is in the realm of HPC already. This is a great starting point on the next feature we need to make available in MsPASS, which is to support external applications to do things that MsPASS alone is not possible to achieve. In this case, we might want to implement the "reduce steps" with OpenMP and then use Spark or Dask to coordinate multiple instance of it on multiple nodes. Or, we may even just call a single MPI program and only use MsPASS to manage the workflow. We definitely need to continue the discussion in our weekly meetings.

**Reply by pavlis** [reply]
Interesting idea.  It seems you are confirming my reading between the lines and projecting serialization of large objects will be an issue is not off base.  In terms of memory use and what does and doesn't need to be serialized there are two different cases here:
1.  For the two outer loops (items 1 and 2 above) the sums are 3d vector field objects, which are always huge.  I think our usarray lower 48 image grid was something like 300x500x800x5=6x10^8 numbers or around 5x10^9 bytes.   For those two steps we have should also think about how the algorithm could be changed to reduce the memory footprint.  Storing even 2 of those things is a memory hog and recall we wanted to do a denser grid run but hit a memory wall.  The obvious way to reduce the memory footprint of summing two grids is to pass the data in blocks to the summer so you only need one copy.   I think we could even do a serial version of that algorithm reading from MongoDB and improve the performance of this program. 
2. The innermost loop in this algorithm is different.  It is effectively like the the block summer I just suggested for summing grids.  That is, the field data are stored in a 4-d data array g.val[i][j][k][l] where the i and j indices at the top of the grid define a map of approximately regular grid points on the earths surface, k is a depth generalized coordinate index, and l is the vector component index.  The inner loop has to fill in k,l values for fixed i and j computed within the loop.   

In a way both algorithms share a common approach which maybe we could generalized - summing large objects in blocks defined by some abstraction of what a "block" means.   

Comments or thoughts on this?
3.

**Reply by wangyinz** [reply]
Yeah, I think that is a feasible approach if we want to fit the algorithm into the MapReduce framework. This is effectively further breaking up the atomic data object to be parallelized. The only caveat here is that we need to make sure that we are not over-complicating things - a threading implementation might be the easier and straightforward solution here. That functionality could also be generalized for MsPASS to support a wide spectrum of external application.

**Comment by pavlis** [comment]
One more possible insight on this problem I had while doing my morning run.  dask's fold and foldby seem to define the general abstraction we need here, but they don't provide the explicit control we probably need due to the large memory problem.   That is:
1.  *dask fold* has the concept of a "partition", which I think could translate to the point I made above about accumulating doing an operation on a large object in pieces.   Perhaps we could think of a way to generalize that by using a partition function to define how a "partition" is defined.  
2. *dask foldby* is similar but it has a key to group data.   I don't think that is as relevant to the specific example here, but I see the value as an abstraction.

I am pretty sure the particular problem with pwmig could be solved pretty easily with explicit threading rather than using dask/spark.  How that would interact with spark/dask or how you could utilize spark/dask to do the threading is way  beyond my expertise.  If I were doing this one up I think I could do a specialized threading version in a few hours because I am nearly certain it could be done with out a mutex, which is usual source of nightmares in multithreaded code.

**Reply by wangyinz** [reply]
Yes, that's exactly the point. It maybe possible to fit the algorithm into the MapReduce framework, but it may not be straightforward. Either way, I think it is also beneficial to explore the use of the "partition" concept in MsPASS, but it doesn't mean we have to implement pwmig with that. There might be other algorithms that are more applicable.

**Comment by pavlis** [comment]
Have been reading about this some more focusing exclusively on dask.   I think I may have a solution I will try, but before jumping in I would like to ask for a critical review by both of you.

For simplicity let me abstract the problem to these three nested loops defined in the following pseudocode:
```
1) Create master 3d grid for global sum
2) for each event
   a) Do some initialization including creating a large 3d object to contain final 3d image
   b) for each plane wave component (can be abstracted as a SeismogramEnsemble)
      i) Compute a 3d grid object for defined by this ensemble's metadata
      ii) for each Seismogram
          A) compute stretching operator
          B) interpolate data 
          C) copy interpolated data into 3D grid object computed at i)
     iii) Sum this grid object into 3D image grid created in a)
     iv) optional save grid object
  c) save 3D image grid just computed
3) sum 3D grid computed in c) to master sum grid
```
Ian, FYI to you only - this does not map exactly into the structure of the old pwmig package you used for you PhD.   in particular, 3) was done as a independent program I called gridstacker.  

This discussion page was created because steps C), iii), and 3) all can be cast in a MapReduce model as as "reduce" operation, BUT all 3 involve very large 3d grid objects on the left hand side of the sum.   Everything I read says that is a bad idea I will need to work around for this to work with any efficiency for reasons noted earlier in this discussion.  

With that background I think a solution with dask alone may be possible with dask delayed.  The reasons is that the documentation says that unless you do something to change it the default scheduler for dask delayed data is threads.   ([Discussed here](https://docs.dask.org/en/latest/setup/single-machine.html))   I thus think a model for doing these large sum is is to use [bag_to_delay](https://docs.dask.org/en/latest/generated/dask.bag.Bag.to_delayed.html#dask.bag.Bag.to_delayed) and then then use the delayed approach to forming these sums.   This is the only thing I've yet found that provides anything like a potentially clean solution for step C) in the pseudocode.  The others have an alternative solution that might be preferable anyway.  That is, by the time we get there in the processing the number of items to handle is not as huge and we could just save all the grid objects to files and drive the sums by a database loop.   The grids have a += operator defined in the C++ library that make the sums always fast an not a limiting calculation.    Step 3), in particular, could be implemented as a pythonization of the old gridstacker program.   

I would like to know if either of you see any dark alleys ahead of me if I pursue his approach.  Also do you have any ideas on how we can test some of these ideas beforehand without implementing the entire monstrous mix of python and C++ it will take to get all of this to work.  I'm thinking maybe some simple test with a large 3d numpy array doing something similar.   Any other suggestions?

**Reply by pavlis** [reply]
Update on this from messing with this early in the morning the past few days in hotel rooms.   I ran some test code and realize that the to_delayed method of a bag provides a generic, useful method to break a large object into pieces (what they call partitions) that can be used to manage memory of large data objects like the pwmig example.   

Here is my example that shows what this does:
```
n=100
m=50
x=np.ndarray(n)
for i in range(n):
    x[i] = float(i)
print(x)
bag=db.range(m,npartitions=5)
```
If I then run a simple function with map followed by to_delayd like this:
```
bag3=bag.map(load_and_scale,x)
delayed_data=bag3.to_delayed()
```
The result (delayed_data above) is a container (type `Delayed`) with 5 entries.  Delayed is iterable and can be used to do lazy computations in controlled chunks.   Here is a useless example using the bag defined earlier:
```
for y in delayed_data:
    z=y.compute()
    print('Chunk return size=',len(z))
```
This is a useful generic way to control memory because the compute method returns a list of the outputs of the chain of functions applied to the bag as lazy computations (load_and_scale for the above example).   In my pwmig example the equivalent to the list z would be a set of data that I could then copy into the large 3D grid.   

The remaining unknown with me is how a reduce/accumulate/fold operator would interact with something like this.  In any case, to me this was an important understanding on how we might use this to provide a more generic partition-based support for large memory algorithms.

Thoughts?
